systemtests/utils/vagrantnode.go

```diff
@@ -15,34 +15,114 @@ limitations under the License.
-	Name    string
-	NodeNum int
+	Name   string
+	client *ssh.Client
+}
+
+//NewVagrantNode intializes a node in vagrant testbed
+func NewVagrantNode(name string, nodeNum int, port, privKeyFile string) (*VagrantNode, error) {
+	var (
+		vnode      *VagrantNode
+		err        error
+		signer     ssh.Signer
+		privateKey []byte
+	)
+
+	if privateKey, err = ioutil.ReadFile(privKeyFile); err != nil {
+		return nil, err
+	}
+
+	if signer, err = ssh.ParsePrivateKey(privateKey); err != nil {
+		return nil, err
+	}
+
+	config := &ssh.ClientConfig{
+		User: "vagrant",
+		Auth: []ssh.AuthMethod{
+			ssh.PublicKeys(signer),
+		},
+	}
+
+	vnode = &VagrantNode{Name: name}
+	if vnode.client, err = ssh.Dial("tcp", fmt.Sprintf("127.0.0.1:%s", port), config); err != nil {
+		return nil, err
+	}
+
+	return vnode, nil
+}
+
+//Cleanup clears the ssh client resources
+func (n *VagrantNode) Cleanup() {
+	n.client.Close()
+}
+
+func newCmdStrWithSource(cmd string) string {
+	// we need to source the environment manually as the ssh package client
+	// doesn't do it automatically (I guess something to do with non interative
+	// mode)
+	return fmt.Sprintf("source /etc/profile.d/envvar.sh && %s", cmd)
 }
 
 // RunCommand runs a shell command in a vagrant node and returns it's exit status
-func (n VagrantNode) RunCommand(cmd string) error {
-	vCmd := &VagrantCommand{ContivNodes: n.NodeNum}
-	return vCmd.Run("ssh", n.Name, "-c", cmd)
+func (n *VagrantNode) RunCommand(cmd string) error {
+	var (
+		s   *ssh.Session
+		err error
+	)
+
+	if s, err = n.client.NewSession(); err != nil {
+		return err
+	}
+	defer s.Close()
+
+	return s.Run(newCmdStrWithSource(cmd))
 }
 
 // RunCommandWithOutput runs a shell command in a vagrant node and returns it's
 // exit status and output
-func (n VagrantNode) RunCommandWithOutput(cmd string) (string, error) {
-	vCmd := &VagrantCommand{ContivNodes: n.NodeNum}
-	output, err := vCmd.RunWithOutput("ssh", n.Name, "-c", cmd)
+func (n *VagrantNode) RunCommandWithOutput(cmd string) (string, error) {
+	var (
+		s   *ssh.Session
+		err error
+	)
+
+	if s, err = n.client.NewSession(); err != nil {
+		return "", err
+	}
+	defer s.Close()
+
+	output, err := s.CombinedOutput(newCmdStrWithSource(cmd))
 	return string(output), err
 }
 
 // RunCommandBackground runs a background command in a vagrant node
-func (n VagrantNode) RunCommandBackground(cmd string) (string, error) {
-	vCmd := &VagrantCommand{ContivNodes: n.NodeNum}
-	output, err := vCmd.RunWithOutput("ssh", n.Name, "-c", cmd)
-	return string(output), err
+func (n *VagrantNode) RunCommandBackground(cmd string) (string, error) {
+	var (
+		s   *ssh.Session
+		err error
+	)
+
+	if s, err = n.client.NewSession(); err != nil {
+		return "", err
+	}
+	defer s.Close()
+
+	// start and forget about the command as user asked to run in background.
+	// The limitation is we/ won't know if it fails though. Not a worry right
+	// now as the test will fail anyways, but might be good to find a better way.
```

Comment:
I wonder if you don't call `s.Wait()`, even in the background, if it leaves zombies lying around?